pkg/operator/starter.go

```diff
@@ -298,6 +299,17 @@ func RunOperator(ctx context.Context, controllerContext *controllercmd.Controlle
 		return err
 	}
 
+	invalidCertsController := metricscontroller.NewMetricsController(
+		"InvalidCertsController",
+		operatorClient,
+		controllerContext.EventRecorder.WithComponentSuffix("invalid-webhook-certs"),
+		"/var/run/configmaps/service-ca-bundle/service-ca.crt",
```

Comment:
can this run if the file does not exist?

Reply:
yes 👍
1.  the underlying configmap is marked as optional.
2. metrics controller is not marking as degraded when an error occurs (i.e. servica CA cannot be found, prometheus will not be reachable, query does not return expected results). Rather, you'll see shakiness during bootstrap like this:
```
surbania@sur-mac ~ % kubectl logs -n openshift-kube-apiserver-operator kube-apiserver-operator-6dd5844588-n78h2 | grep -i invalidcert
I0113 10:51:24.940115       1 base_controller.go:67] Waiting for caches to sync for InvalidCertsController
I0113 10:51:24.946207       1 base_controller.go:73] Caches are synced for InvalidCertsController
I0113 10:51:24.946290       1 base_controller.go:110] Starting #1 worker of InvalidCertsController controller ...
E0113 10:51:54.965934       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp: i/o timeout
E0113 10:52:25.169538       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp: i/o timeout
E0113 10:52:55.187374       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:53:25.188064       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:53:55.188747       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:54:25.189496       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:54:55.190331       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:55:25.191217       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:55:55.192580       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:56:25.193706       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:56:55.194591       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:57:25.195512       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:57:55.196970       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:58:25.198880       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:58:55.202963       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": dial tcp 172.30.45.100:9091: i/o timeout
E0113 10:59:13.336539       1 base_controller.go:272] InvalidCertsController reconciliation failed: empty vector result from query: "sum(apiserver_kube_aggregator_x509_missing_san_total{job=\"apiserver\",namespace=\"default\"}) + sum(apiserver_webhooks_x509_missing_san_total{job=\"apiserver\",namespace=\"default\"})"
E0113 10:59:24.993012       1 base_controller.go:272] InvalidCertsController reconciliation failed: empty vector result from query: "sum(apiserver_kube_aggregator_x509_missing_san_total{job=\"apiserver\",namespace=\"default\"}) + sum(apiserver_webhooks_x509_missing_san_total{job=\"apiserver\",namespace=\"default\"})"
```

Reply:
More formally I extracted the questioned failure mode. We'll see something like this in the logs:
```
I0113 10:50:55.201492       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-kube-apiserver-operator", Name:"kube-apiserver-operator", UID:"1782c25b-da97-4e6c-9e0a-89b3a43e9409", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'PrometheusClientFailed' error instantiating prometheus client: error reading service CA: open /var/run/configmaps/service-ca-bundle/service-ca.crt: no such file or directory
```

and this event will be generated:
```
46m         Warning   PrometheusClientFailed                   deployment/kube-apiserver-operator              error instantiating prometheus client: error reading service CA: open /var/run/configmaps/service-ca-bundle/service-ca.crt: no such file or directory
```

Reply:
And it picks up the configmap later? Does kubelet mount an optional configmap eventually when it exists?

Reply:
I did observe an eventual pickup of the configmap in the logs in the same process:
```
E0113 10:50:55.201492       1 base_controller.go:272] InvalidCertsController reconciliation failed: error instantiating prometheus client: error reading service CA: open /var/run/configmaps/service-ca-bundle/service-ca.crt: no such file or directory
...
E0113 10:51:20.585831       1 base_controller.go:272] InvalidCertsController reconciliation failed: error executing prometheus query: Post "https://thanos-querier.openshift-monitoring.svc:9091/api/v1/query": context canceled
```

But let me dig up some upstream docs, good point.

Reply:
Ah here it is: https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically

> When a mounted ConfigMap is updated, the projected content is eventually updated too. This applies in the case where an optionally referenced ConfigMap comes into existence after a pod has started.